### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by theianrobertson**
Hi @jbrambleDC, came across your package when I was thinking about building something similar, so thanks for that!

There were a couple of things broken when I was poking around with the code (text-based types were only populating `n-1` records for one) and one thing lead to another and I wound up refactoring and adding some testing.  `python -m pytest` runs the tests.

Are you interested in merging these changes up into your repo/package?  If not feel free to close the PR and I'll just keep it on a fork - don't want to step on any toes :).  I've got a couple other things I was thinking of supporting, specifying null/missing rates being a big one that crops up in my day job.

One breaking change I made was switching the calling pattern of the individual type functions to expose them as parameters to the function instead of just passing the `ty` dictionary.  `num_int` for example:

```python
def num_int(self, ty, length):
    a = ty['min']
    b = ty['max']
    return pd.Series(np.random.random_integers(a, b, length))
```

vs.

```python
def num_int(length, minimum=0, maximum=100):
    """Random integers"""
    return pd.Series(np.random.random_integers(minimum, maximum, length))
```

Makes things more explicit imho, and the functions are split out of the class for easier testing.

### CODE REVIEW

**Comment by Gmousse** [comment]
Hi ! I m also interested by your refactor.

I m agreed with your point of view (specially the points about functions and named parameters).
However, creating breaking changes in the API for a refactor make me sad.

I will look this night your code to see if we can create a "compatibility" mode.

**Comment by theianrobertson** [comment]
Good point, and I agree. My reason for swapping for `minimum` and `maximum` was to avoid confusion with built-ins `min` and `max`.  Probably right that other functions should stay with the previous parameter names - I've reverted in 9707cae.  It doesn't break anything to have the parameters as `min` and `max` but it goes against some best practices - if it's a big thing I'm fine swapping back to the originals.

File: simulacrum/__init__.py
Comment (Gmousse) [inline simulacrum/__init__.py]: It could be better to avoid star import everywhere.

I prefer to see `from .dataset import create`
If the user want another function, he can use the complete path to dataset.

File: tests/dataset_test.py
Comment (Gmousse) [inline tests/dataset_test.py]: Same than previously, prefer to explicitly define imports than using star.
`from simulacrum.dataset import create, validate_type_dict, default_coltypes, TYPE_FUNCTIONS`

File: simulacrum/types.py
Comment (Gmousse) [inline simulacrum/types.py]: That's pretty. It's typically my code style. 
I prefer to see these functions than the original contributor's classes.

Well done

File: simulacrum/types.py
Comment (Gmousse) [inline simulacrum/types.py]: Useless sys import

File: simulacrum/types.py
Comment (Gmousse) [inline simulacrum/types.py]: Hum it's a breaking change. Indeed @jbrambleDC originally used min and max as you said in your pull request description...
I don't really care about these parameter names. @jbrambleDC have to choose.

However, A workaround could be to add a high order function (typically a decorator) on num_data and num_int to ensure a compatibility mode on parameters (and print a warning to indicate a deprecation of min and max).
Example:

````python
def ensure_num_compatibility(func):
    """Make sure num_data or num_int will be compatible with minimum and maximum parameters."""
    def compatibility_func(*args, **kwargs):
        compatible_kwargs = dict(kwargs)
        def replace_in_kwargs(parameter_to_replace, new_parameter):
            if parameter_to_replace in kwargs:
                compatible_kwargs[new_parameter] = kwargs[parameter_to_replace]
                del compatible_kwargs[parameter_to_replace]
        replace_in_kwargs('min', 'minimum')
        replace_in_kwargs('max', 'maximum')
        return func(*args, **compatible_kwargs)
    return compatibility_func


@ensure_num_compatibility
def num_int(length, minimum=0, maximum=100):
    """Random integers"""
    return pd.Series(np.random.random_integers(minimum, maximum, length))


num_int(10, min=10, max=22)

````

Comment (Gmousse) [reply simulacrum/types.py]: or even more functional 

````python
def replace_in_kwargs(kwargs, parameter_to_replace, new_parameter):
    """Replace a named parameter in kwargs by another."""
    new_kwargs = dict(kwargs)
    if parameter_to_replace in new_kwargs:
        new_kwargs[new_parameter] = kwargs[parameter_to_replace]
        del new_kwargs[parameter_to_replace]
    return new_kwargs


def ensure_num_compatibility(func):
    """Make sure num_data or num_int will be compatible with minimum and maximum parameters."""
    def compatibility_func(*args, **kwargs):
        compatible_kwargs = replace_in_kwargs(replace_in_kwargs(kwargs, 'min', 'minimum'), 'max', 'maximum')
        return func(*args, **compatible_kwargs)
    return compatibility_func


@ensure_num_compatibility
def num_int(length, minimum=0, maximum=100):
    """Random integers"""
    return pd.Series(np.random.random_integers(minimum, maximum, length))


num_int(10, min=10, max=22)
````

File: simulacrum/types.py
Comment (Gmousse) [inline simulacrum/types.py]: Same than num_data

**Review by Gmousse — CHANGES_REQUESTED** [review]
It's a pretty PR ! When I have contributed to this repository (faker part), I have seen some strange patterns or some misconceptions in the code. You correct a major part of these.
Your functional style make me happy.

When I see this clean codebase, I want to continue to contribute on this project (and I will).

Here I suggest few minor improvements (I don't have the time to do more this night, I will continue to look the code and play with the lib tomorrow) and ore more important about a workaround of your minimum and maximum parameter (I was already faced on parameter rename and breaking changes in some API in other of my work.)

I hope @jbrambleDC will come soon to confirm our direction.

**Comment by Gmousse** [comment]
It could be better to have guideline with a linter configured on this project (flake8, pep8...) because my idle is crying.

I suggest to use tox to fix some rules and to apply unit tests on both python2.7 and 3+.

**Comment by theianrobertson** [comment]
Cool thanks for the feedback - I'd be happy with either option for the min --> minimum.  Agreed-upon docstring style would be good too in addition to some kind of linting guideline - I'd opt for numpy-style if it were up to me, but that's just a personal preference.

Tox I'm less familiar with as I typically use Conda, and I don't think they play nicely together.  Probably would include pinning versions of dependencies too.  I'd be happy to poke around with some of that, but probably want to handle it in separate PRs to avoid this one growing too much.  I'll defer to @jbrambleDC for input on this PR.

**Comment by jbrambleDC** [comment]
Thanks all for the input. This started out as a simple friday night hack more than a year and a half ago when my software engineering abilities were much less than they are now. I am humbled and happy to see people find it interesting enough to contribute to!

I am traveling for the weekend, but I will review as soon as I get the chance. I would most definitely like to get this merged ASAP. 

Thanks to both of you for your contributions.

**Comment by jbrambleDC** [comment]
@theianrobertson Interestingly enough I am traveling to Toronto! Hope its a great time!

File: simulacrum/types.py
Comment (Gmousse) [reply simulacrum/types.py]: Again more clean
````python
from functools import reduce

def replace_in_kwargs(kwargs, parameter_to_replace, new_parameter):
    """Replace a named parameter in kwargs by another."""
    new_kwargs = dict(kwargs)
    if parameter_to_replace in new_kwargs:
        print("WARN The current parameter {} is deprecated. Please consider using {}."
            .format(parameter_to_replace, new_parameter))
        new_kwargs[new_parameter] = kwargs[parameter_to_replace]
        del new_kwargs[parameter_to_replace]
    return new_kwargs


def ensure_num_compatibility(func, **compatibilities):
    """Make sure a function will be compatible parameters change."""
    def compatibility_func(*args, **kwargs):
        compatible_kwargs = reduce(
            lambda current_kwargs, compatibility: replace_in_kwargs(current_kwargs, *compatibility),
            list(zip(compatibilities.keys(), compatibilities.values())),
            kwargs
        )
        return func(*args, **compatible_kwargs)
    return compatibility_func


# add the decorator on each function with deprecated parameter
@ensure_num_compatibility(min='minimum', max='maximum')
def num_int(length, minimum=0, maximum=100):
    """Random integers"""
    return pd.Series(np.random.random_integers(minimum, maximum, length))


num_int(10, min=10, max=22)
````

**Comment by Gmousse** [comment]
@jbrambleDC Thanks !!! I hope to see this.

@theianrobertson Yeah, let's talk about guideline in another PR. I don't have real preferences about linter rules. I only want them to be fixed for all contributors.

Have you an idea about a min -> minimum solution ? The compatibility decorator (not really complex actually, it's just a HOF) can be generalized to handle each parameter modifications (as sd -> stdev or others) in the whole application, throwing a warning to the user to prevent the deprecation.
Moreover it's easy for contributors to simply remove the compatibility decorators when we upgrade to a major version of simulacrum (and then remove deprecated things).

I have modified the code decorator to make it usable with each param modification. Let's see.

I m open to other solutions.

File: simulacrum/__init__.py
Comment (jbrambleDC) [reply simulacrum/__init__.py]: Agree, I prefer to avoid * import here.

File: tests/dataset_test.py
Comment (jbrambleDC) [reply tests/dataset_test.py]: same as above.

File: README.md
Comment (jbrambleDC) [inline README.md]: I like adding this default behavior!

File: simulacrum/dataset.py
Comment (jbrambleDC) [inline simulacrum/dataset.py]: I like the idea of keeping of putting TYPE_FUNCTIONS at the top. However I think I would also like to maintain an Object oriented style for DataSet. Thoughts @Gmousse

File: simulacrum/dataset.py
Comment (jbrambleDC) [inline simulacrum/dataset.py:79]: I like the idea of using a validator function. and unpacking the arguments from a dict to feed to the builder function.  However since the arguments for each data builder in the api are different how are we going to validate that that other keys in the dictionary are correct for that type/function? It seems that is not happening now and can be quite onerous to build one function for that. I would have to think a bit on what a good design pattern would be for this. Also please let me know if I am missing anything?

**Comment by jbrambleDC** [comment]
I've tried to give a prelim reivew here. I havent yet had the chance to test this out yet. I am still familiarizing myself a bit and getting back into the zone of this project.

File: README.md
Comment (Gmousse) [reply README.md]: Yeah that's a good idea to build fast demo of the library or for educational purposes.

File: simulacrum/dataset.py
Comment (Gmousse) [reply simulacrum/dataset.py:79]: I think it could be another PR.

A dictionnary of handlers like that could help:

````python
CHECK_TYPES_HANDLER = {
    "my_databuilder": a checktype function to use directly or another dictionnary containing each parameter type to check.
}

````

File: simulacrum/dataset.py
Comment (Gmousse) [reply simulacrum/dataset.py]: Hum okay. I m not totally agreed with the OO style for DataSet because the object self was barely used in the original version (making the DataSet object a simple collection of static functions).

However the DataSet object usage was originally encapsulated into a function like that:
````python
def create(length, **kwargs):
    x = DataSet(length, **kwargs)
    return x.data
````
Then the original simulacrum API didn't exposed directly the DataSet class.

To conclude the old DataSet write wasn't really using OOP advantages and I don't see why we should to keep this. 

However, if you really really want to use a Class for DataSet, I can write a "best of both worlds" version in order to use an OOP usage for DataSet and a functional one for types. @jbrambleDC

**Comment by Gmousse** [comment]
Thanks @jbrambleDC ! It's really interesting.

File: simulacrum/dataset.py
Comment (jbrambleDC) [reply simulacrum/dataset.py]: thats a good point. So I think the question here is really should we push towards stricter object oriented or functional? From a philosophical standpoint the concept of a DataSet seems like something that should be represented by and object oriented model. However, youre right that the way we are leveraging this now seems to lean functional.

Comment (Gmousse) [reply simulacrum/dataset.py]: To my mind the dataset should not exist. In simulacrum, when you look the code and the purpose, you only have functions to write a complete pandas DataFrame by using some parameters. The dataset have not real consistency because we don t use it correctly and we don t give it to our users. 

If we want to write a real DataSet type class it should be exposed directly and could inherit from pandas DataFrame to have a real object (instead of having DataSet.data in which we store an instance of pandas DataFrame). 

But we have to test some code in order to avoid breaking changes. Tell me. I code try a first version tommorow. 

@jbrambleDC

Comment (Gmousse) [reply simulacrum/dataset.py]: To my mind the dataset should not exist. In simulacrum, when you look the code and the purpose, you only have functions to write a complete pandas DataFrame by using some parameters. The dataset have not real consistency because we don t use it correctly and we don t give it to our users. 

If we want to write a real DataSet type class it should be exposed directly and could inherit from pandas DataFrame to have a real object (instead of having DataSet.data in which we store an instance of pandas DataFrame). 

But we have to test some code in order to avoid breaking changes. Tell me. I can 
code try a first version tommorow. 

@jbrambleDC

Comment (jbrambleDC) [reply simulacrum/dataset.py]: You make a convincing argument. If you want to make some changes and so long as they are testable, I will merge. We should also let @theianrobertson chime in. My other request is can we add some handlers so that legacy uses will still function? If not can we make sure it is clearly documented in the README.md Make sure we also update the Changelog and the setup.cfg so we can cut a new version once tested and merged.

Comment (Gmousse) [reply simulacrum/dataset.py]: I will look more preciesly in the code and try tok make some tries tomorrow. Before to make modifications, I want to be sure that we will break nothing. 

@theianrobertson have you any proposal abput this discussion ?

Comment (theianrobertson) [reply simulacrum/dataset.py]: Yeah I was thinking along the same lines as @Gmousse - the docs are using `sm.create`, which can be called unchanged here, other than the `min`/`max`.  With current functionality the only interaction is on creation of the object, and I'm not sure there's anything that *needs* to be in a class.  I could see keeping it as a class if you had other steps when setting up the data, like:

```python
import simulacrum as sm

test = {'entries': {'type': 'exp', 'lam': 0.5},
        'names': {'type': 'name'},
        'salaries': {'type': 'norm', 'mean': 55000, 'sd': 20000}}

res = sm.dataset.DataSet(100, coltypes=test)
res.set_null_rate(column='names', rate=0.3)
res.set_null_rate(column='salaries', rate=0.5)
print(res.data)
```

But then you could also expose that kind of functionality with something like `res = sm.create(100, coltypes=test, null_rates={'names': 0.3, 'salaries': 0.5})` too.

Long story short I like the simplicity of de-classing it.

File: simulacrum/dataset.py
Comment (theianrobertson) [reply simulacrum/dataset.py:79]: Might actually be able to just leverage the base `TypeError` that would be raised:

```python
>>> import simulacrum as sm
>>>
>>> test_good = {'entries': {'type': 'exp', 'lam': 0.5}}
>>> res_good = sm.create(10, coltypes=test_good)
>>>
>>> test_bad = {'entries': {'type': 'exp', 'bad_key': 100}}
>>> res_bad = sm.create(10, coltypes=test_bad)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ian/Documents/github/simulacrum/simulacrum/dataset.py", line 64, in create
    series_res[col] = data_builder(length, **type_dict)
TypeError: exp_data() got an unexpected keyword argument 'bad_key'
>>>
>>> test_bad_type = {'entries': {'type': 'norm', 'mean': '0', 'sd': 1}}
>>> res_bad = sm.create(10, coltypes=test_bad_type)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/ian/Documents/github/simulacrum/simulacrum/dataset.py", line 64, in create
    series_res[col] = data_builder(length, **type_dict)
  File "/Users/ian/Documents/github/simulacrum/simulacrum/types.py", line 35, in norm_data
    return pd.Series(np.random.normal(mean, sd, length))
  File "mtrand.pyx", line 1643, in mtrand.RandomState.normal (numpy/random/mtrand/mtrand.c:20599)
TypeError: must be real number, not str
```

I think those are decently good error messages to raise to a user, at least the former.  For the latter, if you wanted a better message, you could always raise a specific message in `norm_data` like:

```python
def norm_data(length, mean=0, sd=1):
    """Normal distribution data"""
    try:
        return pd.Series(np.random.normal(mean, sd, length))
    except TypeError:
        raise TypeError('mean ({}) and sd ({}) must both be numeric!'.format(mean, sd))
```

Thoughts?  I do agree that any handling that you want to add could be in a separate PR.

Comment (Gmousse) [reply simulacrum/dataset.py:79]: Yep it makes sense.

File: simulacrum/dataset.py
Comment (Gmousse) [reply simulacrum/dataset.py]: Ok ! It's maybe better to keep your current code (because it doesn't change the api).
If we want to make a class api, the PR will never close.
Let's keep the discussion about a class another day.

**Comment by Gmousse** [comment]
@theianrobertson Have you commit the required modifications to merge (like parameters compatibility) ? I m little busy since the last 2 weeks but I can take few minutes to look your last commits.

@jbrambleDC It could be interesting to quickly validate all the PR and merge it if everything is ok (after ian last required modifications).

**Pull request closed without merging**
